rir/src/interpreter/interp.c

```diff
@@ -310,21 +310,33 @@ SEXP rirCallTrampoline(RCNTXT* cntxt, Code* code, SEXP env, unsigned nargs,
 static SEXP rirCallClosure(SEXP call, SEXP env, SEXP callee, SEXP actuals,
                            unsigned nargs, Context* ctx) {
 
-    SEXP body = BODY(callee);
-    Function* fun = sexp2function(body);
+    SEXP vtableStore = BODY(callee);
+    DispatchTable* vtable = sexp2dispatchTable(vtableStore);
+    SEXP funStore = vtable->entry[0];
+    Function* fun = sexp2function(funStore);
 
     static bool optimizing = false;
 
+    // NOTE(mhyee): The introduction of a dispatch table for different function
+    // versions changes many things. This function handles a closure call, and
+    // will optimize the closure function and update the linked list of
+    // functions, so that subsequent closure calls pick the most optimized
+    // version.
+    // This is no longer true with the dispatch table, where the first function
+    // is always selected, unless specified by the call instruction.
+    // For now, to keep the changes simpler, we'll just optimize the function
+    // once and add it to the dispatch table. But the call will still select
+    // the original, unoptimized version.
+
     if (!optimizing &&
         !fun->next
-         /* currently there is a bug if we reoptimize a function twice:
```

Comment:
Looks like `clang-format` fixed the indentation.

As a side note, hopefully this bug will be fixed when we implement deoptimization between versions.